src/_pytest/config/__init__.py

```diff
@@ -392,40 +393,48 @@ def parse_hookimpl_opts(self, plugin: _PluggyPlugin, name: str):
         if name == "pytest_plugins":
             return
 
-        method = getattr(plugin, name)
         opts = super().parse_hookimpl_opts(plugin, name)
+        if opts is not None:
+            return opts
 
+        method = getattr(plugin, name)
         # Consider only actual functions for hooks (#3775).
         if not inspect.isroutine(method):
             return
-
         # Collect unmarked hooks as long as they have the `pytest_' prefix.
-        if opts is None and name.startswith("pytest_"):
-            opts = {}
-        if opts is not None:
-            # TODO: DeprecationWarning, people should use hookimpl
-            # https://github.com/pytest-dev/pytest/issues/4562
-            known_marks = {m.name for m in getattr(method, "pytestmark", [])}
 
-            for name in ("tryfirst", "trylast", "optionalhook", "hookwrapper"):
-                opts.setdefault(name, hasattr(method, name) or name in known_marks)
+        known_marks = {m.name for m in getattr(method, "pytestmark", [])}
+        opts = {
+            name: hasattr(method, name) or name in known_marks
+            for name in ("tryfirst", "trylast", "optionalhook", "hookwrapper")
+        }
+        if any(opts.values()):
```

Comment:
Curious: would `if opts:` work the same way? Seems like this won't warn for this case:

```python
pytest_runtest_call.tryfirst = False
```

Which I understand is unorthodox, however seems like we want to warn in that case anyway?

Reply:
We have to use none then to distinguish between default and missing, will fix

Reply:
implemented, as wed like to keep hooks by name